### DISCUSSION TITLE
mqtt_node keeping large amount of connection details in memory

### DISCUSSION THREAD

**Discussion by johanrhodin** [rabbitmq/rabbitmq-server#2688] [Other] [answered]
As reported in [rabbitmq-users thread "High memory consumption (and message queue) in mqtt_node"](https://groups.google.com/g/rabbitmq-users/c/hiKKiqyweV4) we see RabbitMQ consuming more and more memory and never draining the Erlang mailbox. 

We've created a repo that recreates the issue with a high likelyhood by creating a connection, channel, consumer and queue per connection. One can tune the connection to say 10k, 20k, etc until the situation is triggered. The repo is here: [mqtt-connections](https://github.com/84codes/mqtt-connections)

In addition to the screenshots attached in rabbitmq-users mail thread this is what the stack trace looks like
![mqtt_node_stacktrace](https://user-images.githubusercontent.com/460369/102545514-f88bd200-407b-11eb-9cf0-68fde78c902c.png)

**Comment by carlhoerberg** [comment]
run `node churn.js mqtt://guest:guest@localhost 1` and see how the `ra_log_ets` grows indefinitely

**Comment by kjnilsson** [comment]
What WAL size are you using?

See: https://www.rabbitmq.com/mqtt.html#consensus

On Thu, 17 Dec 2020 at 21:37, Carl Hörberg <notifications@github.com> wrote:

> run node churn.js mqtt://guest:guest@localhost 1 and see how the
> ra_log_ets grows indefinitely
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/rabbitmq/rabbitmq-server/discussions/2688#discussioncomment-220278>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAJAHFDMZZEF35UC7FWALL3SVJ225ANCNFSM4VAIPAXQ>
> .
>


-- 
*Karl Nilsson*

**Comment by carlhoerberg** [comment]
tried with 64 MB, no difference

**Comment by carlhoerberg** [comment]
rabbitmq config, to make high MQTT connection count possible: 
```
[
  {credentials_obfuscation, [{iterations, 1}, {hash, 'sha256'}, {cipher, 'aes_cbc128'}]},
  {rabbitmq_mqtt, [
    {tcp_listen_options, [
      {backlog, 4096},
      {sndbuf, 4096},
      {recbuf, 4096}
    ]}
  ]},
  {rabbitmq_management, [
    {rates_mode, none},
    {disable_stats, true},
    {enable_queue_totals, true}
  ]}
].
```
then run: 
`node app.js mqtt://guest:guest@localhost 10000`, when they're up, close the application so that all connections closes. Now monitor the inbox of the `mqtt_node` process, it takes a very long time before it has reaches zero, if you continue to open/close connection it will grow indefinitely, eating GB after GB of RAM.

The stack trace indicates that it's mostly stuck in this function: https://github.com/rabbitmq/rabbitmq-server/blob/e3bbdfe6dfcd75eb685c8672cf7c619bc3796348/deps/rabbitmq_mqtt/src/mqtt_machine.erl#L72-L83

**Reply by kjnilsson** [reply]
ok so that is the function that is handling the down messages that arise from shutting each connection, I mean it isn't super optimised for sure. do you log at debug level? does it behave any better if you don't?

**Reply by kjnilsson** [reply]
Ok I've done a first-pass optimisation of this function if you feel like testing it: https://github.com/rabbitmq/rabbitmq-server/tree/mqtt-machine-opt

**Reply by kjnilsson** [reply]
hah a bit too quick, just pushed a fix

**Comment by carlhoerberg** [comment]
I do see a difference with a lower `wal_max_size_bytes`, the each mqtt_node ets table is smaller, but there are many of them instead. Total seems to be the same.

**Comment by carlhoerberg** [comment]
Attaching the [crash.log.gz](https://github.com/rabbitmq/rabbitmq-server/files/5712619/crash.log.gz)

**Comment by carlhoerberg** [comment]
The multiple mqtt_node ets tables seems to be cleared up when the inbox for `mqtt_node` is drained (which takes a long time even on an idle machine with no connections, after opened and closed 10 000 connections), but the `mqtt_process` was still using a lot of memory until the process was manually GCed.

**Comment by carlhoerberg** [comment]
Now, when restarting the server after opened the 10000 connections and then closed them, always results in a failed boot: 

```
[error] <0.18031.1> CRASH REPORT Process <0.18031.1> with 0 neighbours exited with reason: {timeout,{gen_statem,
@carl-xps13-fedora'},trigger_election,5000]}} in application_master:init/4 line 138
[info] <0.44.0> Application rabbitmq_mqtt exited with reason: {timeout,{gen_statem,call,[{mqtt_node,'rabbit@carl
election,5000]}}
[error] <0.17838.1> BOOT FAILED
[error] <0.17838.1> ===========
[info] <0.44.0> Application rabbitmq_mqtt exited with reason: {timeout,{gen_statem,call,[{mqtt_node,'rabbit@carl
election,5000]}}
[error] <0.17838.1> Error during startup: {error,
[error] <0.17838.1>                        {rabbitmq_mqtt,
[error] <0.17838.1>                         {bad_return,
[error] <0.17838.1>                          {{rabbit_mqtt,start,[normal,[]]},
[error] <0.17838.1>                           {'EXIT',
[error] <0.17838.1>                            {timeout,
[error] <0.17838.1>                             {gen_statem,call,
[error] <0.17838.1>                              [{mqtt_node,'rabbit@carl-xps13-fedora'},
[error] <0.17838.1>                               trigger_election,5000]}}}}}}}
```

Nothing but deleting the quorum directory seems to get it moving.

**Comment by kjnilsson** [comment]
I will take a closer look next week. What kind of disks are you using?

On Thu, 17 Dec 2020 at 23:37, Carl Hörberg <notifications@github.com> wrote:

> Now, when restarting the server after opened the 10000 connections and
> then closed them, always results in a failed boot:
>
> [error] <0.18031.1> CRASH REPORT Process <0.18031.1> with 0 neighbours exited with reason: {timeout,{gen_statem,
> @carl-xps13-fedora'},trigger_election,5000]}} in application_master:init/4 line 138
> [info] <0.44.0> Application rabbitmq_mqtt exited with reason: {timeout,{gen_statem,call,[{mqtt_node,'rabbit@carl
> election,5000]}}
> [error] <0.17838.1> BOOT FAILED
> [error] <0.17838.1> ===========
> [info] <0.44.0> Application rabbitmq_mqtt exited with reason: {timeout,{gen_statem,call,[{mqtt_node,'rabbit@carl
> election,5000]}}
> [error] <0.17838.1> Error during startup: {error,
> [error] <0.17838.1>                        {rabbitmq_mqtt,
> [error] <0.17838.1>                         {bad_return,
> [error] <0.17838.1>                          {{rabbit_mqtt,start,[normal,[]]},
> [error] <0.17838.1>                           {'EXIT',
> [error] <0.17838.1>                            {timeout,
> [error] <0.17838.1>                             {gen_statem,call,
> [error] <0.17838.1>                              [{mqtt_node,'rabbit@carl-xps13-fedora'},
> [error] <0.17838.1>                               trigger_election,5000]}}}}}}}
>
> Nothing but deleting the quorum directory seems to get it moving.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/rabbitmq/rabbitmq-server/discussions/2688#discussioncomment-220483>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAJAHFCTI3C6JXDY65OUCMTSVKI3DANCNFSM4VAIPAXQ>
> .
>
-- 
*Karl Nilsson*

**Comment by carlhoerberg** [comment]
We've tried with AWS GP2, local SSD and RAM disk, no difference, no real
disk usage anyway, just a single core very busy

**Comment by carlhoerberg** [comment]
Awesome! Will try it out in a bit

**Reply by kjnilsson** [reply]
A test I ran for this code that adds 10k connections then processes dows for each of them went from 38s to 4s - still not super quick but may be enough to keep up for your use-case

**Reply by kjnilsson** [reply]
I also have an approach that can bring the same test down to a few ms but it will require a bit more work and testing.

**Reply by carlhoerberg** [reply]
I can't replicate the problem any more with this build, it seems to have fixed the trigger_election boot error too! This is great!

**Reply by kjnilsson** [reply]
ok thanks for testing! I will prepare a patch with an optimised mqtt machine, I think I will do the full, proper optimisation and check if there are any other rough edges in the machine.

**Reply by michaelklishin** [reply]
@kjnilsson so there are more changes coming? @carlhoerberg it would be great if you could help us verify more versions before we adopt it for `3.8.10` (which is almost ready to ship and is mostly delayed because of a very PTO-heavy season)

**Reply by kjnilsson** [reply]
@michaelklishin yes there will be more changes, I can get it down from 4s to a few ms for 10000 downs so I think it's worth it

**Reply by carlhoerberg** [reply]
sure, but "verify more versions", of rabbitmq or of tests cases that replicate the problem?

**Reply by kjnilsson** [reply]
I've created a PR now: https://github.com/rabbitmq/rabbitmq-server/pull/2692

**Comment by carlhoerberg** [comment]
Just tested it, my initial trials shows a huge difference!

**Comment by michaelklishin** [comment, marked as answer]
https://github.com/rabbitmq/rabbitmq-server/pull/2692 seems to address this.

**Comment by johanrhodin** [comment]
We'll take a build with #2692 in it for a few spins.
